staging/src/k8s.io/apiserver/pkg/storage/tests/cacher_test.go

```diff
@@ -773,4 +782,121 @@ func TestCacherListerWatcherPagination(t *testing.T) {
+	rv, err := v.ParseResourceVersion(fooCreated.ResourceVersion)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	startVersion := strconv.Itoa(int(rv))
+
+	tests := []struct {
+		timeout            time.Duration
+		expected           bool
+		allowWatchBookmark bool
+	}{
+		{ // test old client won't get Bookmark event
+			timeout:            0,
+			expected:           false,
+			allowWatchBookmark: false,
+		},
+		{
+			timeout:            2*time.Second + 10*time.Millisecond,
+			expected:           true,
+			allowWatchBookmark: true,
+		},
+	}
+
+	for _, c := range tests {
+		pred := storage.Everything
+		pred.AllowWatchBookmarks = c.allowWatchBookmark
+		ctx, _ := context.WithTimeout(context.Background(), c.timeout)
+		watcher, err := cacher.Watch(ctx, "pods/ns/foo", startVersion, pred)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		// Create events of other pods
+		updatePod(t, etcdStorage, makeTestPod("foo-whatever"), nil)
+
+		// Now wait for Bookmark event
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok && c.expected {
+				t.Errorf("Unexpected object watched (no objects)")
+			}
+			if c.expected && event.Type != watch.Bookmark {
+				t.Errorf("Unexpected object watched %#v", event)
+			}
+		case <-time.After(time.Second * 3):
+			if c.expected {
+				t.Errorf("Unexpected object watched (timeout)")
+			}
+		}
+		watcher.Stop()
+	}
+}
+
+func TestWatchBookmarksWithCorrectResourceVersion(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.WatchBookmark, true)()
+
+	server, etcdStorage := newEtcdTestStorage(t, etcdtest.PathPrefix())
+	defer server.Terminate(t)
+	cacher, v := newTestCacher(etcdStorage, 10)
+	defer cacher.Stop()
+
+	pred := storage.Everything
+	pred.AllowWatchBookmarks = true
+	ctx, _ := context.WithTimeout(context.Background(), 2*time.Second)
+	watcher, err := cacher.WatchList(ctx, "pods/ns", "0", pred)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer watcher.Stop()
+
+	done := make(chan struct{})
+	go func() {
+		for i := 0; i < 100; i++ {
+			pod := fmt.Sprintf("foo-%d", i)
+			err := createPod(etcdStorage, makeTestPod(pod))
+			if err != nil {
+				t.Fatalf("failed to create pod %v", pod)
+			}
+			time.Sleep(time.Second / 100)
+		}
+		close(done)
+	}()
+	defer func() {
+		<-done
+	}()
+
+	bookmarkResourceVersion := uint64(0)
+	lastObservedResourceVersion := uint64(0)
+	for event := range watcher.ResultChan() {
+		rv, err := v.ObjectResourceVersion(event.Object)
+		if err != nil {
+			t.Fatalf("failed to parse resourceVersion from %#v", event)
+		}
+		if event.Type == watch.Bookmark {
+			bookmarkResourceVersion = rv
+			if bookmarkResourceVersion < lastObservedResourceVersion {
+				t.Fatalf("Unexpected bookmark resourceVersion %v less than observed %v)", bookmarkResourceVersion, lastObservedResourceVersion)
+			}
+		} else {
+			if rv <= bookmarkResourceVersion {
```

Comment:
I think this can be simplified - every non-bookmark event has to have RV greater than anything before.
So you can just switch to:
```
if rv <= lastObservedResourceVersion {
  ...
}
```

Note that then you also don't need the bookmarkResourceVersion variable, which will simplify this whole test.